### PR TITLE
feat: accurately estimate lootrun beacon position

### PR DESCRIPTION
**Pull request by kristofbolyai**
Superseds https://github.com/Wynntils/Artemis/pull/1906

This PR has code that allows 100% correct location prediction, so we only have to assign  to beacons to already known challange locations, which makes it much more accurate as well.

image

### CODE REVIEW

**Comment by magicus** [comment]
So we're using particles now? :-o

**Comment by kristofbolyai** [comment]
> So we're using particles now? :-o

Particles and beacons :) Double fun.

File: common/src/main/java/com/wynntils/features/combat/CustomLootrunBeaconsFeature.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+package com.wynntils.features.combat;
+
+import com.wynntils.core.consumers.features.Feature;
+import com.wynntils.core.persisted.Persisted;
+import com.wynntils.core.persisted.config.Category;
+import com.wynntils.core.persisted.config.Config;
+import com.wynntils.core.persisted.config.ConfigCategory;
+
+@ConfigCategory(Category.COMBAT)
+public class CustomLootrunBeaconsFeature extends Feature {
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/combat/CustomLootrunBeaconsFeature.java:14]: What happens if you disable this feature? I found where you check the config about hiding beacons, but I assume that you will only draw the new beacons if this feature is enabled, right? I just can't figure out where that rendering is happening.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/combat/CustomLootrunBeaconsFeature.java:14]: `LootrunBeaconMarkerProvider#isEnabled`

**Comment by kristofbolyai** [comment]
Did some changes to colors and added text color texts to make this much more visible during runs.

image
(the blue is a tiny bit dark but it is acceptable imo)

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java [hunk 2]
```diff
@@ -23,13 +24,14 @@
...
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.phys.Vec3;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
 import org.apache.commons.compress.utils.Lists;
 
 public class BeaconModel extends Model {
     // Amount of armor stands above each other to consider this a beacon
-    // (A beacon typically has around 17 in total)
-    private static final int VERIFICATION_ENTITY_COUNT = 6;
+    // (A beacon always has 34, we wait for all of them)
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:33]: Height is 256, not 128..

**Comment by kristofbolyai** [comment]
I am sorry for making this PR so huge. I had to change 2 different models: Markers, Lootrun and add a whole new one, Particles. I hope it it still clear enough to review though.

Also, I would like to say that the particle model seems really great for use on Wynn, we should use it for a lot of cool stuff :)

File: common/src/main/java/com/wynntils/features/combat/CustomLootrunBeaconsFeature.java (hunk 1 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/features/combat/CustomLootrunBeaconsFeature.java:14]: I see. I was looking for something in the map package; I forgot about how you designed the new providers. Thanks!

File: common/src/main/java/com/wynntils/utils/colors/CommonColors.java [hunk 3]
```diff
@@ -16,6 +16,7 @@ public final class CommonColors {
     public static final CustomColor BROWN = CustomColor.fromInt(0x563100).withAlpha(255);
     public static final CustomColor PURPLE = CustomColor.fromInt(0xb200ff).withAlpha(255);
     public static final CustomColor CYAN = CustomColor.fromInt(0x438e82).withAlpha(255);
+    public static final CustomColor AQUA = CustomColor.fromInt(0x00ffff).withAlpha(255);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/utils/colors/CommonColors.java:19]: I have just added:
```
    public static final CustomColor AQUA = CustomColor.fromInt(0x00cccc).withAlpha(255);
```

in my mapdata branch, for how the province place labels where colored. That is apparently a darker shade of aqua. Do we try to converge on one value, or should I rename mine DARK_AQUA?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/colors/CommonColors.java:19]: I am happy with how close this version is to the Wynn AQUA color, so can you perhaps rename yours?

**Comment by magicus** [comment]
This looks great. It was easy to review even if it was large. I was about to say the same about the particle model. I'm going to add some logging that prints to the console whenever we see particles, and then I'm going to run around to see if I can catch them at some interesting places. :-)

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java (hunk 2 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:33]: I'm not sure what you mean by this..?

**Review by magicus — APPROVED** [review]

File: common/src/main/java/com/wynntils/utils/colors/CommonColors.java (hunk 3 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/utils/colors/CommonColors.java:19]: Okidoki.

File: common/src/main/java/com/wynntils/models/beacons/BeaconModel.java (hunk 2 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:33]: Last time you added that comment by doing 128/7.5, but I misslead you. The height is 256.

Comment (magicus) [reply common/src/main/java/com/wynntils/models/beacons/BeaconModel.java:33]: Aha.

**Pull request merged**